sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
+      // First time we need to allow for spilling
+      spill = Some(SpillableColumnarBatch(cached.get,
+        SpillPriorities.ACTIVE_ON_DECK_PRIORITY,
+        spillCallback))
+      // Putting data in a SpillableColumnarBatch takes ownership of it.
+      cached = None
+    }
+    cached.foreach(_.close())
+    cached = None
+  }
+
+  override def close(): Unit = synchronized {
+    cached.foreach(_.close())
+    cached = None
+    spill.foreach(_.close())
+    spill = None
+  }
+}
+
+object JoinGathererImpl {
+
+  /**
+   * Calculate the row size in bits for a fixed width schema. If a type is encountered that is
+   * not fixed width, or is not known a None is returned.
+   */
+  def fixedWidthRowSizeBits(dts: Seq[DataType]): Option[Int] =
+    sumRowSizesBits(dts, nullValueCalc = false)
+
+  /**
+   * Calculate the null row size for a given schema in bits. If an unexpected type is enountered
+   * an exception is thrown
+   */
+  def nullRowSizeBits(dts: Seq[DataType]): Int =
+    sumRowSizesBits(dts, nullValueCalc = true).get
+
+
+  /**
+   * Sum the row sizes for each data type passed in. If any one of the sizes is not available
+   * the entire result is considered to not be available. If nullValueCalc is true a result is
+   * guaranteed to be returned or an exception thrown.
+   */
+  private def sumRowSizesBits(dts: Seq[DataType], nullValueCalc: Boolean): Option[Int] = {
+    val allOptions = dts.map(calcRowSizeBits(_, nullValueCalc))
+    if (allOptions.exists(_.isEmpty)) {
+      None
+    } else {
+      Some(allOptions.map(_.get).sum + 1)
+    }
+  }
+
+  /**
+   * Calculate the row bit size for the given data type. If nullValueCalc is false
+   * then variable width types and unexpected types will result in a None being returned.
+   * If it is true variable width types will have a value returned that corresponds to a
+   * null, and unknown types will throw an exception.
+   */
+  private def calcRowSizeBits(dt: DataType, nullValueCalc: Boolean): Option[Int] = dt match {
+    case StructType(fields) =>
+      sumRowSizesBits(fields.map(_.dataType), nullValueCalc)
+    case dt: DecimalType if dt.precision > DType.DECIMAL64_MAX_PRECISION =>
+      if (nullValueCalc) {
+        throw new IllegalArgumentException(s"Found an unsupported type $dt")
+      } else {
+        None
+      }
+    case _: NumericType | DateType | TimestampType | BooleanType | NullType =>
+      Some(GpuColumnVector.getNonNestedRapidsType(dt).getSizeInBytes * 8 + 1)
+    case StringType | BinaryType | ArrayType(_, _) if nullValueCalc =>
+      // Single offset value and a validity value
+      Some((DType.INT32.getSizeInBytes * 8) + 1)
+    case x if nullValueCalc =>
+      throw new IllegalArgumentException(s"Found an unsupported type $x")
+    case _ => None
+  }
+}
+
+/**
+ * JoinGatherer for a single map/table
+ */
+class JoinGathererImpl(
+    // TODO need a way to spill/cache the GatherMap
+    private val gatherMap: GatherMap,
```

Comment:
Technically if you don't make it a val, except for a case class, you are accessing a closure value instead of an actual val. They might be implemented the same way, but I wanted to be explicit.